modules/admin/admin-impl/src/main/java/com/enonic/xp/admin/impl/rest/resource/project/icon/ProjectIconResource.java

```diff
@@ -0,0 +1,192 @@
+package com.enonic.xp.admin.impl.rest.resource.project.icon;
+
+import java.awt.image.BufferedImage;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Paths;
+import java.util.Iterator;
+
+import javax.annotation.security.RolesAllowed;
+import javax.imageio.ImageIO;
+import javax.imageio.ImageWriter;
+import javax.ws.rs.DefaultValue;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.CacheControl;
+import javax.ws.rs.core.Response;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+import com.google.common.hash.HashCode;
+import com.google.common.hash.Hashing;
+import com.google.common.io.ByteSource;
+
+import com.enonic.xp.admin.impl.rest.resource.content.ResolvedImage;
+import com.enonic.xp.attachment.Attachment;
+import com.enonic.xp.home.HomeDir;
+import com.enonic.xp.image.ImageHelper;
+import com.enonic.xp.jaxrs.JaxRsComponent;
+import com.enonic.xp.project.Project;
+import com.enonic.xp.project.ProjectName;
+import com.enonic.xp.project.ProjectService;
+import com.enonic.xp.repository.RepositoryService;
+import com.enonic.xp.security.RoleKeys;
+import com.enonic.xp.util.HexEncoder;
+import com.enonic.xp.util.ImmutableFilesHelper;
+
+import static com.enonic.xp.admin.impl.rest.resource.ResourceConstants.REST_ROOT;
+import static com.google.common.base.Strings.isNullOrEmpty;
+
+@Path(REST_ROOT + "project/icon")
+@Produces("image/*")
+@RolesAllowed({RoleKeys.ADMIN_LOGIN_ID, RoleKeys.ADMIN_ID})
+@Component(immediate = true, property = "group=admin")
+public final class ProjectIconResource
```

Comment:
This is a controller class which does waaay too much.

Let see how to split it up.